### PR TITLE
partition methods on AssetExecutionContext

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Adds `partition_key_range_for_asset_key` to `AssetExecutionContext`. Also does some minor reorganizing of partition methods on `StepExecutionContext` to support passing AssetKeys rather than input names

To Do:
- [ ] update tests to use new methods, instead of deprecated ones
- [ ] determine if we need to add a `partition_keys_for_asset_key` method to support materializing multiple static partitions

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16480** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16480" target="_blank">Graphite</a> 
    * **PR #16596** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16596" target="_blank">Graphite</a> 
      * **PR #16493** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16493" target="_blank">Graphite</a> 
        * **PR #16598** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16598" target="_blank">Graphite</a> 
          * **PR #16507** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16507" target="_blank">Graphite</a> 
            * **PR #16625** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16625" target="_blank">Graphite</a>  👈
    * **PR #16486** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16486" target="_blank">Graphite</a> 
      * **PR #16487** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16487" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16625?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/execution/context/compute.py [hunk 1]
```diff
@@ -1481,9 +1485,56 @@ def asset_check_spec(self) -> AssetCheckSpec:
         return self._op_execution_context.asset_check_spec
 
     @public
-    def partition_key_range_for_asset_key(self, asset_key: AssetKey) -> PartitionKeyRange:
-        """TODO - implement in stacked pr."""
-        pass
+    def partition_key_range_for_asset_key(
+        self, asset_key: CoercibleToAssetKey, is_dependency: bool = False
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/execution/context/compute.py:1489]: still not super happy with the `is_dependency` solution to the self dependent asset problem. here's my comment from the first PR exploring this:

--- 
the is_dependency bool param here is to handle the self-dependent partition case. Basically when you have a self-dependent partition, if you said `context.partition_key_for_asset_key("my_asset")` it's unclear if you want the partition key of my_asset as a dependency, or as is being currently materialized. I'm handling this (for now at least) by allowing users to specify `is_dependency` to indicate they want it loaded as a dependency.

In all other cases, the function should be able to figure out the correct partition key to load given the current asset

I'm not set on this solution, so if you have any concerns or other ideas let me know

**Comment by jamiedemaria** [comment]
Discussing and aligning on partition method APIs here https://github.com/dagster-io/internal/discussions/6704. then will implement in this PR

**Review by alangenfeld — CHANGES_REQUESTED** [review]
whats the overlap with this and https://github.com/dagster-io/dagster/pull/16480? 

back to your queue

**Comment by jamiedemaria** [comment]
closing for now. I think determining the partition method APIs will be best done in a discussion.

**Pull request closed without merging**